source/extensions/filters/listener/local_ratelimit/local_ratelimit.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include "envoy/event/file_event.h"
+#include "envoy/event/timer.h"
+#include "envoy/extensions/filters/listener/local_ratelimit/v3/local_ratelimit.pb.h"
+#include "envoy/network/filter.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "source/common/common/logger.h"
+#include "source/common/runtime/runtime_protos.h"
+#include "source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace LocalRateLimit {
+
+/**
+ * All local rate limit stats. @see stats_macros.h
+ */
+#define ALL_LOCAL_RATE_LIMIT_STATS(COUNTER) COUNTER(rate_limited)
+
+/**
+ * Struct definition for all local rate limit stats. @see stats_macros.h
+ */
+struct LocalRateLimitStats {
+  ALL_LOCAL_RATE_LIMIT_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+using LocalRateLimiterImplSharedPtr =
+    std::shared_ptr<Filters::Common::LocalRateLimit::LocalRateLimiterImpl>;
+
+/**
+ * Configuration shared across all connections. Must be thread safe.
+ */
+class Config : Logger::Loggable<Logger::Id::filter> {
```

Comment:
It's actually filter config. I saw the convention is on the same file with filter. Should I rename to `FilterConfig` to avoid confusion?

Reply:
Normally configuration and filter implementations are split into dedicated files. I would expect the config class to be in the config.h/cc files together with the factory.

Reply:
Factory initializes the filter, and filter needs the config information. 
That is, if we put factory and config into `config.h/cc`, we have circular dependency between config module and filter module, which fails the bazel check. 

I'm unsure the we have precedence putting config with factory. 
We have one case in http land `jwt_authn`, which put config into its own file.
However it needs to have another own file to handle factory to avoid circular dependency.

I could do three pairs of .h/.cpp if we really want to. 

```
 git grep "class FilterConfig" -- *.h
source/extensions/filters/http/alternate_protocols_cache/filter.h:17:class FilterConfig {
source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.h:37:class FilterConfig {
source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.h:67:class FilterConfigImpl : public FilterConfig {
source/extensions/filters/http/bandwidth_limit/bandwidth_limit.h:61:class FilterConfig : public ::Envoy::Router::RouteSpecificFilterConfig {
source/extensions/filters/http/custom_response/config.h:30:class FilterConfig : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/ext_authz/ext_authz.h:54:class FilterConfig {
source/extensions/filters/http/ext_authz/ext_authz.h:242:class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/ext_proc/ext_proc.h:68:class FilterConfig {
source/extensions/filters/http/ext_proc/ext_proc.h:108:class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/grpc_http1_reverse_bridge/filter.h:69:class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/jwt_authn/filter_config.h:41:class FilterConfig {
source/extensions/filters/http/jwt_authn/filter_config.h:62:class FilterConfigImpl : public Logger::Loggable<Logger::Id::jwt>,
source/extensions/filters/http/local_ratelimit/local_ratelimit.h:70:class FilterConfig : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/lua/lua_filter.h:375:class FilterConfig : Logger::Loggable<Logger::Id::lua> {
source/extensions/filters/http/lua/lua_filter.h:414:class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/oauth2/filter.h:128:class FilterConfig {
source/extensions/filters/http/ratelimit/ratelimit.h:41:class FilterConfig {
source/extensions/filters/http/ratelimit/ratelimit.h:118:class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/tap/tap_filter.h:37:class FilterConfig {
source/extensions/filters/http/tap/tap_filter.h:57:class FilterConfigImpl : public FilterConfig, public Extensions::Common::Tap::ExtensionConfigBase {
source/extensions/filters/http/wasm/wasm_filter.h:24:class FilterConfig : Logger::Loggable<Logger::Id::wasm> {
source/extensions/filters/network/wasm/wasm_filter.h:24:class FilterConfig : Logger::Loggable<Logger::Id::wasm> {
```

Reply:
I think the question here is : Do you need this separate class`class config`? Also, looks like the `config` class doesn't use synchronization primitive or use TLS (i.e., RCU pattern). So, I am not sure thread safe is guaranteed here when this class  is not read-only. (canCreateConnection() is  fine as its underlying `requestAllowedHelper` use atomic operation)

I think it could just work if we remove config class and its members/methods are moved into to `class Filter`? The threading safe is also naturally achieved.  WDYT?

Reply:
On second thought, it is good to have this `config` class as it is, since its members are read only and rate limiter is not shared.

Maybe rename it to 'FilterConfig' to be consistent with other places in code base.

Reply:
done. thanks!